rust/tests/optimize_test.rs

```diff
@@ -0,0 +1,91 @@
+#[cfg(feature = "datafusion-ext")]
+mod optiize {
```

Comment:
I think you may want a few more tests. Here are some properties we likely want to test for:

 1. Optimize is idempotent. If you run twice, the second time it won't write any new files.
 2. Optimize [bin packs](https://en.wikipedia.org/wiki/Bin_packing_problem). For example, with a max file size of 100MB, files of size 70MB, 100MB, 30MB will turn into two files of 100MB, regardless of the order they show up in the log.
 3. Optimize fails if a concurrent writer overwrites the tables. It might be able to succeed if a concurrent writer appends to the table, at the very least in the case where the append happens in a different partition.

Reply:
I changed the implementation to sort files to be optimized to ensure that it is idempotent. Added a couple of tests to validate that behavior too. For item 3 I think it will have to wait until we have a generalized pattern for non-append writers.